Sources/LeafKit/LeafAST.swift

```diff
@@ -0,0 +1,127 @@
+public typealias ResolvedDocument = LeafAST
+
+// Internal combination of rawAST = nil indicates no external references
+// rawAST non-nil & flat = false : unresolved with external references, and public AST is not flat
+// rawAST non-nil & flat = true : resolved with external references, and public AST is flat
+public struct LeafAST: Hashable {
```

Comment:
Would it be possible to make this internal? Anything we can make internal is much easier to maintain and extend after 4.0.0 release goes out.

Reply:
Much obliged!

Currently it's not possible to make internal because I've left LeafCache's definition of storage to ResolvedDocument to prevent breaking potential , so aliasing LeafAST requires it to be public. If you don't think that's an issue (I can't see anything on GitHub to suggest anyone's rolling their own cache backend but it's certainly possible someone has) I'd definitely to drop ResolvedDocument as a type altogether.

I do think there's an advantage to having a public-level interface to it though, so that users can actually query the AST's behavior. One of my goals is to make it feasible to have API-level configurability of Vapor, so that the same infrastructure can service multiple theming variants, and it would be necessary for an application to be able to query the ASTs directly to ensure they adhere to required methods at that point to be usable, which would then involve additional observers for reporting that an AST expects to be provided certain contextual data or key names. Happy for thoughts on that though.

If nothing else, I don't see any problem with making it internal for now while it's still in flux and then opening it up later once things stabilize and are trusted to not need change.

Reply:
(Also note that while it's currently public, the only publicly usable methods are all getters only and ones I would not expect to ever change).